### PR TITLE
Don't let the non-Enso types float around the Enso interpreter!

### LINKED ISSUE #7552: Remove list of hard-coded type names from `TypeWithKind.java`
For now I can just extend it with `EnsoObject` interface, but we should use `processEnv.getTypes()` to check the real types and not just names.

_Originally posted by @JaroslavTulach in https://github.com/enso-org/enso/pull/7544#discussion_r1289976750_

### LINKED ISSUE #9477: Oracle GraalVM fails on AssertionError: no root node
This happens when (dry-running) some benchmarks. Locally, this can be reproduced with:
```
sbt:std-benchmarks> benchOnly org.enso.benchmarks.generated.Map_Error_Benchmark_Slice_report_warning.Map_Id_All_Errors
```

Link to a corresponding failure on the CI: org.enso.benchmarks.generated.Map_Error_Benchmark_Slice_report_warning.Map_Id_All_Errors

The error is:
```
[error] java.lang.AssertionError: no root node for ArraySliceGen.TypesLibraryExports.Cached@4d523257
[error] with section: null
[error] with root nodes: null
[error] 	at org.enso.runtime/org.enso.interpreter.runtime.EnsoContext.reportSlowContextAccess(EnsoContext.java:243)
[error] 	at org.enso.runtime/org.enso.interpreter.runtime.EnsoContext.get(EnsoContext.java:218)
[error] 	at org.enso.runtime/org.enso.interpreter.runtime.data.vector.ArraySlice.getType(ArraySlice.java:185)
[error] 	at org.enso.runtime/org.enso.interpreter.runtime.data.vector.ArraySliceGen$TypesLibraryExports$Cached.getType(ArraySliceGen.java:101)
[error] 	at org.enso.runtime/org.enso.interpreter.runtime.library.dispatch.TypesLibraryGen$CachedDispatch.getType(TypesLibraryGen.java:529)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeMethodNodeGen.execute(InvokeMethodNodeGen.java:192)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNode.invokeDynamicSymbol(InvokeCallableNode.java:268)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNodeGen.execute(InvokeCallableNodeGen.java:149)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:97)
[error] 	at org.enso.runtime/org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callDirect(OptimizedCallTarget.java:535)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:94)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.thunk.ThunkExecutorNode.doCached(ThunkExecutorNode.java:69)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.thunk.ThunkExecutorNodeGen.executeThunk(ThunkExecutorNodeGen.java:123)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.argument.ArgumentSorterNode.executeArguments(ArgumentSorterNode.java:84)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.argument.ArgumentSorterNode.execute(ArgumentSorterNode.java:100)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.InvokeFunctionNode.invokeCached(InvokeFunctionNode.java:107)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.InvokeFunctionNodeGen.execute(InvokeFunctionNodeGen.java:81)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNode.invokeFunction(InvokeCallableNode.java:167)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNodeGen.execute(InvokeCallableNodeGen.java:125)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:97)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:54)
[error] 	at org.enso.runtime/org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callDirect(OptimizedCallTarget.java:535)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:94)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ExecuteCallNode.callDirect(ExecuteCallNode.java:94)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ExecuteCallNodeGen.executeCall(ExecuteCallNodeGen.java:86)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.SimpleCallOptimiserNode.executeDispatch(SimpleCallOptimiserNode.java:56)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.CurryNode.doCall(CurryNode.java:161)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.CurryNode.execute(CurryNode.java:107)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.InvokeFunctionNode.invokeCached(InvokeFunctionNode.java:116)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.InvokeFunctionNodeGen.execute(InvokeFunctionNodeGen.java:81)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNode.invokeFunction(InvokeCallableNode.java:167)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.InvokeCallableNodeGen.execute(InvokeCallableNodeGen.java:125)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:97)
[error] 	at org.enso.runtime/org.enso.interpreter.node.scope.AssignmentNodeGen.executeGeneric_generic1(AssignmentNodeGen.java:78)
[error] 	at org.enso.runtime/org.enso.interpreter.node.scope.AssignmentNodeGen.executeGeneric(AssignmentNodeGen.java:55)
[error] 	at org.enso.runtime/org.enso.interpreter.node.scope.AssignmentNodeGen.executeVoid(AssignmentNodeGen.java:98)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:52)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:54)
[error] 	at org.enso.runtime/org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callDirect(OptimizedCallTarget.java:535)
[error] 	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedDirectCallNode.call(OptimizedDirectCallNode.java:94)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ExecuteCallNode.callDirect(ExecuteCallNode.java:94)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.ExecuteCallNodeGen.executeCall(ExecuteCallNodeGen.java:86)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.SimpleCallOptimiserNode.executeDispatch(SimpleCallOptimiserNode.java:56)
[error] 	at org.enso.runtime/org.enso.interpreter.node.callable.dispatch.CurryNode.doCall(CurryNode.java:161)
...
```

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Failure described in https://github.com/enso-org/enso/pull/9462#issuecomment-2027928285 discovered that some of our builtins are returning wrong type. Only `long`, `double`, `boolean` and generic `TruffleObject` types are expected in the Enso interpreter. Returning any other type (for example `java.lang.String`) leads to _increased polymorphism_ and slowdown (or higher memory consumption) of the interpreter as the Truffle AST gets richer to handle more types.

This PR modifies the annotation processor to print a warning when a builtin's _return type is suspicious_. Failures are analyzed and fixed. Remaining cases where a generic `java.lang.Object` is being returned [are annotated](https://github.com/enso-org/enso/pull/9584/commits/4375728327e509dd0b43a48709f2af9702908e0d) by `@SuppressWarnings("generic-enso-builtin-type")` annotation.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides
- All code has been tested:
  - [x] Compiler warnings analyzed and resolved

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Relates to https://github.com/enso-org/enso/issues/1239#issuecomment-802742723

**Comment by JaroslavTulach** [comment]
Here are the [warnings about suspicious return types](https://github.com/enso-org/enso/actions/runs/8489263653/job/23259164045?pr=9584#step:8:850):
```
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:54:1:  Suspicious return type: java.io.OutputStream
    public OutputStream outputStream(Object opts, EnsoContext ctx) throws IOException {
                        ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:65:1:  Suspicious return type: java.io.InputStream
    public InputStream inputStream(Object opts, EnsoContext ctx) throws IOException {
                       ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:153:1:  Suspicious return type: java.util.Set<java.nio.file.attribute.PosixFilePermission>
    public Set<PosixFilePermission> getPosixPermissions() throws IOException {
                                    ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java:159:1:  Suspicious return type: java.lang.Object
    public Object getParent() {
                  ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ManagedResource.java:63:1:  Suspicious return type: java.lang.Object
    public Object take(EnsoContext context) {
                  ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Ref.java:33:1:  Suspicious return type: java.lang.Object
    public Object getValue() {
                  ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeHelpers.java:83:1:  Suspicious return type: java.lang.Object
    public static Object vectorFromFunction(
                         ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:46:1:  Suspicious return type: java.lang.Object
    public Object getValue() {
                  ^
 engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:155:1:  Suspicious return type: java.lang.Object
    public static Object set(
```
@hubertp, @Akirathan let's discuss what do  we want to do with them.

**Comment by JaroslavTulach** [comment]
As a result of bcf69a3702a3da2916e91448209ba9a88a9990ca there is now [AssertionError failure in tests](https://github.com/enso-org/enso/actions/runs/8489263653/job/23259164045?pr=9584#step:8:9675). It should be gone once we merge #9462 - e.g. latest develop in.

**Comment by JaroslavTulach** [comment]
[This](https://github.com/enso-org/enso/blame/c42e6d19dadebab0bfe852902766f4033f4004ef/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java#L134) is wrong, @hubertp.

https://github.com/enso-org/enso/blob/c42e6d19dadebab0bfe852902766f4033f4004ef/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java#L134

results in 
```ruby
  Object execute(EnsoFile self) {
    EnsoContext context = EnsoContext.get(this);
    try {
      return context
          .asGuestValue(self.getCreationTime());
```
e.g. a `TruffleObject` gets wrapped in `asGuestValue`. The only way it can work is that `asGuestValue` does no wrapping of `TruffleObject`, but than the `@Builtin.ReturningGuestObject` annotation is unnecessary.

Since a98f0b7 we error on such unnecessary conversion.

File: lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/TypeWithKind.java [hunk 1]
```diff
@@ -23,42 +22,46 @@ static TypeWithKind createFromTpe(String tpeName) {
...
+  public static boolean isValidGuestType(ProcessingEnvironment processingEnv, TypeMirror type) {
+    return switch (type.getKind()) {
+      case BOOLEAN, LONG, DOUBLE -> true;
+      case VOID -> true;
+      case DECLARED -> isTruffleObject(processingEnv, type);
+      default -> false;
+    };
+  }
+
+  static boolean isTruffleObject(ProcessingEnvironment processingEnv, TypeMirror type) {
```
Comment (hubertp) [inline lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/TypeWithKind.java:55]: :+1:

**Review by hubertp — APPROVED** [review]
Thanks, always nice to get rid of that hardcoded list.

**Comment by JaroslavTulach** [comment]
Here is the current set of [suspicious types](https://github.com/enso-org/enso/actions/runs/8506305757/job/23296285611?pr=9584). Let's:
- not print the warning when there is `@Builtin.ReturningGuestObject`
- let's allow to supress the warning in case return type is `Object`.

**Pull request merged**